services/s3/src/it/java/software/amazon/awssdk/services/s3/checksum/AsyncHttpChecksumIntegrationTest.java

```diff
@@ -291,4 +295,82 @@ public void putObject_with_bufferCreatedFromZeroCapacityByteBuffer() {
         assertThat(interceptor.responseValidation()).isEqualTo(ChecksumValidation.VALIDATED);
         assertThat(response).isEqualTo("");
     }
+
+    private static Stream<Arguments> getObjectChecksumValidationParams() {
+        return Stream.of(Arguments.of(true, ChecksumAlgorithm.CRC32, ChecksumMode.ENABLED),
+                         Arguments.of(true, null, ChecksumMode.ENABLED),
+                         Arguments.of(true, ChecksumAlgorithm.CRC32, null),
+                         Arguments.of(true, null, null),
+                         Arguments.of(false, ChecksumAlgorithm.CRC32, ChecksumMode.ENABLED),
+                         Arguments.of(false, null, ChecksumMode.ENABLED),
+                         Arguments.of(false, ChecksumAlgorithm.CRC32, null),
+                         Arguments.of(false, null, null));
+    }
+
+    @ParameterizedTest
+    @MethodSource("getObjectChecksumValidationParams")
+    public void testGetObjectChecksumValidation(boolean checksumValidationEnabled, ChecksumAlgorithm checksumAlgorithm,
+                                                ChecksumMode checksumMode) {
+        S3AsyncClient s3Async = s3AsyncClientBuilder().overrideConfiguration(o -> o.addExecutionInterceptor(interceptor))
+                                                 .serviceConfiguration(S3Configuration.builder()
+                                                                                      .checksumValidationEnabled(checksumValidationEnabled)
+                                                                                      .build())
+                                                 .build();
+
+        s3Async.putObject(PutObjectRequest.builder()
+                                     .bucket(BUCKET)
+                                     .key(KEY)
+                                     .checksumAlgorithm(checksumAlgorithm)
+                                     .build(), AsyncRequestBody.fromString("Hello world")).join();
+
+        s3Async.getObject(GetObjectRequest.builder()
+                                     .bucket(BUCKET)
+                                     .key(KEY)
+                                     .checksumMode(checksumMode)
+                                     .build(), AsyncResponseTransformer.toBytes()).join();
+
+        validateChecksumValidation(checksumValidationEnabled, checksumAlgorithm, checksumMode);
+        interceptor.reset();
+    }
+
+    private void validateChecksumValidation(boolean checksumValidationEnabled, ChecksumAlgorithm checksumAlgorithm,
+                                            ChecksumMode checksumMode) {
+        if (checksumValidationEnabled) {
```

Comment:
nit: nested if-else is a bit hard to read. Can we break it down to separate methods?